8.8/apache-buster/Dockerfile

```diff
@@ -57,17 +56,23 @@ RUN { \
 		echo 'opcache.fast_shutdown=1'; \
 	} > /usr/local/etc/php/conf.d/opcache-recommended.ini
 
-WORKDIR /var/www/html
-
 # https://www.drupal.org/node/3060/release
 ENV DRUPAL_VERSION 8.8.8
-ENV DRUPAL_MD5 c8265f0228ca0306de7b2755e4731a3d
 
+# https://github.com/drupal/drupal/blob/9.0.1/composer.lock#L4052-L4053
+COPY --from=composer:1.10 /usr/bin/composer /usr/local/bin/
 RUN set -eux; \
-	curl -fSL "https://ftp.drupal.org/files/projects/drupal-${DRUPAL_VERSION}.tar.gz" -o drupal.tar.gz; \
-	echo "${DRUPAL_MD5} *drupal.tar.gz" | md5sum -c -; \
-	tar -xz --strip-components=1 -f drupal.tar.gz; \
-	rm drupal.tar.gz; \
-	chown -R www-data:www-data sites modules themes
+	export COMPOSER_HOME="$(mktemp -d)"; \
+	cd /var/www/; \
+	# composer won't install to a non-empty dir: 'Project directory "/var/www/" is not empty.'
+	rmdir /var/www/html; \
+	composer create-project --no-install "drupal/recommended-project:$DRUPAL_VERSION" ./; \
+	sed -i 's!web/!html/!g' composer.json; \
```

Comment:
Should we maybe *not* do this for 9.0+?  (I understand why we would do it for 8.x to keep backwards compatibility, but if it's the "recommended" layout, perhaps we should follow it for the new versions and update our Apache configuration instead?)

Reply:
I was thinking to always do it so that we don't have the weird transition for users where they get confused that their mount for `/var/www/html/modules` no longer works (https://github.com/docker-library/docs/tree/2d2443ed9d71dafa0f328b546c39f6562603ce7d/drupal#volumes)

(honestly the "recommended" layout should have some way to choose the target subdirectory)

I guess we could do both by installing to web and making html a symlink?

Reply:
Letting it install where they "recommend" and having our own symlink seems sane IMO -- gives us a good balance of following their official layout and not breaking users.

Reply:
You can provide it like:
```
composer create-project --no-install "drupal/recommended-project:$DRUPAL_VERSION"  /var/www/html
```
per the documentation:
https://www.drupal.org/docs/develop/using-composer/using-composer-to-install-drupal-and-manage-dependencies#s-create-a-project

Reply:
or am I misunderstanding the problem?

Reply:
oh I see: https://github.com/drupal/recommended-project/blob/e0e5638a868be7b0334de2404912fd57bec03b96/composer.json#L35-L49

that's annoying... :(

Reply:
Yeah, a little -- when we do `composer create-project ...  /var/www/html` it creates `/var/www/html/web` which is the actual "serve these files" directory.  We were hoping for an easy to way to just make `.../web` _be_ `/var/www/html` so we don't have to change the `php` base image's Apache configuration.

Reply:
I mean you could override the composer.json file I guess....

Reply:
Right, I see, that's what you're doing.

Reply:
as far as I know, the "root" `composer.json` file is "customizable" by the user (i.e. you would not use it to load modules or themes or whatever). So I think it's fine to modify it when you build it personally... it seems better than having a symlink that I might not know about....

Reply:
it might be weird if someone mounts the `/var/www` folder and expects the `web` subdirectory to work, but I would just document that they need to change `web` to `html` in _their_ `composer.json`

Reply:
What it sounds like is that we're maybe looking at the problem in the wrong way -- what if we instead install to something like `/opt/drupal`, and then make `/var/www/html` a symlink out to `/opt/drupal/web`?  Then we get the best of both hacks, without all the weird edge cases:

- bind mounts into `/var/www/html/...` will still work
- no Apache configuration changes required
- users can still provide all of `/var/www/html` or even use `/opt/drupal` themselves to manage (and the standard "recommended" layout will Just Work)

Maybe there's even an official upstream-recommended location for a "system" install (rather than inventing `/opt/drupal` ourselves)?

Reply:
That seems reasonable to me. That means `/opt/drupal` will point to it's subdirectory `web` and `/var/html` wont contain the `composer.json` file in the first place. :)

I do not know of a system install location convention, so whatever you think is best. Is there a convention in Docker? maybe `/app`?

The documentation should probably be updated to use/mount `/opt/drupal` (and I suppose that should be the `WORKDIR` ?) instead of `/var/www`.

It would be nice if I could extend the image, and do something like:
```
 composer require 'drupal/paragraphs:^1.12'
```
and have it install the Paragraphs module as is described in the documentation:
https://www.drupal.org/project/paragraphs/releases/8.x-1.12

Reply:
Great example usage! After some tweaks that I'll push up shorty, this will work just fine.

```console
$ docker build -
FROM drupal:8.9-new
RUN composer require 'drupal/paragraphs:^1.12'

Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM drupal:8.9-new
 ---> 82e0b578fbc4
Step 2/2 : RUN composer require 'drupal/paragraphs:^1.12'
 ---> Running in d7ad66c57f7b
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 2 installs, 0 updates, 0 removals
As there is no 'unzip' command installed zip files are being unpacked using the PHP zip extension.
This may cause invalid reports of corrupted archives. Besides, any UNIX permissions (e.g. executable) defined in the archives will be lost.
Installing 'unzip' may remediate them.
  - Installing drupal/entity_reference_revisions (1.8.0): Downloading (100%)         
  - Installing drupal/paragraphs (1.12.0): Downloading (100%)         
drupal/paragraphs suggests installing drupal/entity_browser (Recommended for an improved user experience when using the Paragraphs library module)
Writing lock file
Generating autoload files
23 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
Removing intermediate container d7ad66c57f7b
 ---> fc0923aa982b
Successfully built fc0923aa982b
```